File: ruby/003. largest prime factor.rb
# Largest prime factor
# https://projecteuler.net/problem=3

target = 600851475143
target_s = 13195

def isPrime(n)
    if n == 2
        return true
    end

    if n < 2 || n % 2 == 0
        return false
    end

    for i in 3..(Math.sqrt(n)).to_i
        if i % 2 == 0
            next
        end
        if n % i == 0
            return false
        end
    end
    return true
end

def largestPrimeFactor(n)
    if n < 2
        return nil
    end
    if isPrime(n)
        return n
    end

    (Math.sqrt(n)).to_i.downto(1) do |i|
        if n % i != 0
            next
        end
        if isPrime(i)
            return i
        end
    end
    return nil
end

puts(largestPrimeFactor(target))            

File: c/003_largest_prime_factor.c
// Largest prime factor
// https://projecteuler.net/problem=3

#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#include "lib/prime.h"

const long long big_int = 600851475143;

long long largestPrimeFactor(long long number) {
    if (number == 0 || number == 1) {
        return 0;
    }
    if (isPrime(number)) {
        return number;
    }
    const long long root = sqrt(number);
    long long primeFactors[1024];
    long index = 0;
    for (int i = 2; i <= root; i++) {
        if (number % i != 0) {
            continue;
        }
        if (isPrime(i)) {
            primeFactors[index] = i;
            index++;
        }
        if (isPrime(number / i)) {
            primeFactors[index] = number / i;
            index++;
        }
    }
    long long largest = -1;
    for (int i = 0; i < index; i++) {
        if (primeFactors[i] > largest) {
            largest = primeFactors[i];
        }
    }
    return largest;
}

int main(void) {
    long long lpm = largestPrimeFactor(big_int);
    printf("%lld\n", lpm);
    return 0;
}
